### DISCUSSION TITLE
How can a quarto extension trigger multiple renders for a quarto document automatically based on different profile environments?

### DISCUSSION THREAD

**Discussion by coatless** [quarto-dev/quarto-cli#5836] [Q&A] [answered]
### Description

I'm exploring using Quarto's [environment profiles](https://quarto.org/docs/projects/profiles.html) to generate multiple versions of a document.

From the documentation, if we have two environment files, e.g. `_quarto-advanced.yml` and `_quarto-production.yml`, then we can render a single document using:

```sh
# multiple profiles on command line
quarto render --profile advanced,production
```

This makes an assumption that one of the profiles specifies an `output-dir:` value that differs from the working directory.

My main question is can I have a quarto extension _automatically_ render the file under multiple profiles if they exist _without_ having the user issue the `quarto render` command? That Is, I want them to be able to press a button in either RStudio or VS Code and the output just **appears**.

From there, a couple of side questions are:

1. Can the name of the rendered quarto document be modified during the save?
2. Is there a way to introduce a simplified class name for the code cell? For instance, if we have:

````
```{.content-visible when-profile="production"}
# Some code
```
````

We would want to use a new class name such as `.env-prod` to decrease the **amount** of typing.

````
```{.env-prod}
# Some code
```
````

**Comment by cscheid** [comment, marked as answer]
> From there, a couple of side questions are:
> 
> 1. Can the name of the rendered quarto document be modified during the save?
> 2. Is there a way to introduce a simplified class name for the code cell? For instance, if we have:
> 
> ````
> ```{.content-visible when-profile="production"}
> # Some code
> ```
> ````
> 
> We would want to use a new class name such as `.env-prod` to decrease the **amount** of typing.
> 
> ````
> ```{.env-prod}
> # Some code
> ```
> ````

As you have noticed, this is currently not documented very well, but we're working on improving it. With that said, let me explain to you here how to achieve the outcome of your second question.

What we're going to do is we'll write a Lua filter that converts `.env-prod` to a conditional block.

In 1.4, conditional blocks are declared using `.content-visible`, but are implemented as a "Custom AST" node, a special kind of node in the Lua filter that extends the nodes provided by Pandoc (like callouts and panel tabsets). Custom AST nodes are recognized and parsed before user filters. So if your user filter emits a div with `.content-visible`, it's too late for quarto to process it. Instead, you have to call the quarto API explicitly. The filter would look something like this: (NB: I haven't tested it!)

```lua
function Div(div)
  if not div.classes:includes(".env-prod") then
    return nil
  end
  local behavior = "content-visible" -- or "content-hidden", as necessary
  local condition = {
    {"when-profile", "production"},
    -- other conditions
  }
  return quarto.ConditionalBlock({
    node = div, -- this is the div containing your content
    behavior = behavior,
    condition = condition
  })
end
```

Your question highlights two quarto shortcomings we hope to address:

1. We should offer an API to process both markdown or "raw Pandoc AST" into custom AST nodes, so the div syntax works in user filters
2. We should document the APIs for constructing Custom AST nodes.

**Reply by coatless** [reply]
Thanks for the details on the custom parsing. I modified it a little bit and incorporated it into an extension geared to replace one of my existing RMarkdown-specific packages. 

<https://github.com/coatless/quarto-assign>

The only other showstopper is being able to have _all_ profiles rendered when the button is clicked. Is there a good way to specify this inside of the `_extension.yml`?

**Reply by cscheid** [reply]
> The only other showstopper is being able to have all profiles rendered when the button is clicked

That's a question for the IDE folks, either [RStudio](https://github.com/rstudio/rstudio) or [VS Code](https://github.com/quarto-dev/quarto).
